2-Draft-Accepted/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,150 @@
+---
+RFC: RFCxxxx
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 0.1
+Area: Security
+Comments Due: July 31st, 2019
+Plan to implement: Yes, PS7
+---
+
+# Simplify and secure use of secrets in scripts
+
+Advanced scripts that touch many systems require multiple secrets and
+types of secrets particularly when orchestrating across different clouds.
+Best practice is to not hard code any secrets in scripts, but currently
+this requires custom code on different platforms to handle this securely.
+
+This RFC proposes new cmdlets to make this simpler and secure.
+
+## Motivation
+
+> As a PowerShell script author,
+> I can securely use multiple secrets,
+> so that I can automate complex orchestration across multiple remote resources.
+
+## High Level Design
+
+Secrets are stored securely in a local vault.
+The local vault is expected to only allow access to the user who owns that
+vault.
+Secrets required to access remote vaults are stored in the local vault and used by the secrets management
+cmdlets to retrieve remote secrets.
+
+`User Context` --> `Local Vault` --> `SecretsVaultExtension` --> `Remote Vault`
+
+## User Experience
+
+Registering and using remote secrets:
+
+```powershell
+# Install the Az.KeyVault module to be used in this example.
+# This module doesn't conform to the extensions model so would need to be
+# updated to actually work.
+Install-Module Az.KeyVault
+
+# In this example, we explicitly register this extension
+Register-SecretsVaultExtension -Name Azure -Cmdlet Get-AzKeyVaultSecret `
```

Comment:
This hasn't been called out yet, but I believe that there are two problems with this cmdlet name:
#1 - It's got a plural (it should be Register-SecretVaultExtension, not Register-SecretsVaultExtension)
#2 - It's a mouthful.  While switching the verb to Add- will chop off a lot of that mouthful, I believe something more succinct would be better.

How about:
* Add-SecretProvider (probably clear to avg users, probably a little weird to seasoned PS users unless you make a provider)
* Add-SecretServer
* Add-SecretSource


Beyond that I don't see anything about disambiguation or rank.  Can I only have one source for secrets?  If I have N, how do I manage which ones get priority?

Reply:
I was OK with the plural here, and don't find it to be the same as a pluralized noun. It's registering a Secrets Vault, not a registering a Secret Vault (which sounds like it's hidden somewhere). That said, I won't lose any sleep either way with that part of the name.

I also prefer `Register` over `Add` here -- it appropriately and accurately describes what it does.

Reply:
Sticking with `Register` for the Vault, dropping `Extension` suffix.  As @KirkMunro noted, the noun is `SecretsVault` which itself is singular.